x-pack/plugins/security_solution/public/common/components/exceptions/add_exception_modal/index.tsx

```diff
@@ -0,0 +1,337 @@
+  >([]);
+  const [fetchOrCreateListError, setFetchOrCreateListError] = useState(false);
+  const [, dispatchToaster] = useStateToaster();
+  const { loading: isSignalIndexLoading, signalIndexName } = useSignalIndex();
+
+  const [{ isLoading: indexPatternLoading, indexPatterns }] = useFetchIndexPatterns(
+    signalIndexName !== null ? [signalIndexName] : []
+  );
+
+  const onError = useCallback(
+    (error: Error) => {
+      errorToToaster({ title: i18n.ADD_EXCEPTION_ERROR, error, dispatchToaster });
+      onCancel();
+    },
+    [dispatchToaster, onCancel]
+  );
+  const onSuccess = useCallback(() => {
+    displaySuccessToast(i18n.ADD_EXCEPTION_SUCCESS, dispatchToaster);
+    onConfirm();
+  }, [dispatchToaster, onConfirm]);
+
+  const [{ isLoading: addExceptionIsLoading }, addOrUpdateExceptionItems] = useAddOrUpdateException(
+    {
+      http,
+      onSuccess,
+      onError,
+    }
+  );
+
+  const handleBuilderOnChange = useCallback(
+    ({
+      exceptionItems,
+    }: {
+      exceptionItems: Array<ExceptionListItemSchema | CreateExceptionListItemSchema>;
+    }) => {
+      setExceptionItemsToAdd(exceptionItems);
+    },
+    [setExceptionItemsToAdd]
+  );
+
+  const onFetchOrCreateExceptionListError = useCallback(
+    (error: Error) => {
+      setFetchOrCreateListError(true);
+    },
+    [setFetchOrCreateListError]
+  );
+  const [isLoadingExceptionList, ruleExceptionList] = useFetchOrCreateRuleExceptionList({
+    http,
+    ruleId,
+    exceptionListType,
+    onError: onFetchOrCreateExceptionListError,
+  });
+
+  useEffect(() => {
+    if (alertData !== undefined && exceptionListType === 'endpoint' && ruleExceptionList) {
+      setExceptionItemsToAdd(
+        defaultEndpointExceptionItems(
+          exceptionListType,
+          ruleExceptionList.list_id,
+          ruleName,
+          alertData
+        )
+      );
+    }
+  }, [alertData, exceptionListType, ruleExceptionList, ruleName, setExceptionItemsToAdd]);
+
+  useEffect(() => {
+    if (indexPatternLoading === false && isSignalIndexLoading === false) {
+      setShouldDisableBulkClose(
+        entryHasListType(exceptionItemsToAdd) ||
+          entryHasNonEcsType(exceptionItemsToAdd, indexPatterns)
+      );
+    }
+  }, [
+    setShouldDisableBulkClose,
+    exceptionItemsToAdd,
+    indexPatternLoading,
+    isSignalIndexLoading,
+    indexPatterns,
+  ]);
+
+  const onCommentChange = useCallback(
+    (value: string) => {
+      setComment(value);
+    },
+    [setComment]
+  );
+
+  const onCloseAlertCheckboxChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setShouldCloseAlert(event.currentTarget.checked);
+    },
+    [setShouldCloseAlert]
+  );
+
+  const onBulkCloseAlertCheckboxChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setShouldBulkCloseAlert(event.currentTarget.checked);
+    },
+    [setShouldBulkCloseAlert]
+  );
+
+  const enrichExceptionItems = useCallback(() => {
```

Comment:
I might move some of these functions out into a `helpers.ts` file so you can unit test them easily.

Reply:
We'll look into moving the entire `enrichExceptionItems` function to the helpers file in the follow up PR.